Dynamic Mode Fixes: Take II

Cenrus opened the pull request:
https://github.com/tgstation/tgstation/pull/46132
https://github.com/tgstation/tgstation/pull/47100
https://github.com/tgstation/tgstation/pull/47341
https://github.com/tgstation/tgstation/pull/47351
https://github.com/tgstation/tgstation/pull/47395
https://github.com/tgstation/tgstation/pull/47491
https://github.com/tgstation/tgstation/pull/47763
https://github.com/tgstation/tgstation/pull/48202
https://github.com/tgstation/tgstation/pull/54345
https://github.com/tgstation/tgstation/pull/44026

First PR had some issues that I couldn't understand, so I did a clean cherrypick and now the fixes work.

:cl:TheChosenEvilOne, Skoglol, Cenrus, and others
fix: Dynamic mode fixes
fix: Revolution ending dynamic earlier than expected
add: Clockwork Cult and Abductors have been implemented into dynamic mode
add: The revolution gamemode will no longer be selected if there are no head of staff players
add: Dynamic added to secret rotation
/:cl:

TheChosenEvilOne committed:
Config comments

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
oops

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

kriskog committed:
More dynamic fixes (#47351)

* Fixes roundstart() runtime, adds more logging.

* one more log line.

* And fixed those lines.

* Fixed highpop runtime, toned down traitors a bit.

* Missing )

* fixed log

* Fixes midround candidate lists.

* right list

* midrounds respect protected jobs

* Err try again but not really

AnturK committed:
Fixes dynamic stalling

kriskog committed:
Even more dynamic fixes. (#47395)

* scaling calculation again.

* No more warnings for antag_cap-less modes

* makes revs work again

* scaling exits earlier of not threat left

* initializes antag_cap again.

kriskog committed:
Adds a solo abductor option to traitor panel.

kriskog committed:
Dynamic: Improved threat logging, no triple refunds (#48202)

* Improved threat logging, no triple refunds

* Update code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm

Antag rolls respect antag_flag_override

Co-Authored-By: TheChosenEvilOne <34602646+TheChosenEvilOne@users.noreply.github.com>

TheChosenEvilOne committed:
Fixes dynamic rev ruleset going bonkers when first assigned is invalid. (#48457)

* Fix rev ruleset behaviour when first assigned is invalid.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Why was this not done originally?

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

kriskog committed:
Fixes dynamic revs runtime

kriskog committed:
fixes for non-carbon current!

kriskog committed:
Fixes cloning message

kriskog committed:
Dynamic can only roll revs once

IndieanaJones committed:
Fixes Dynamic Abductor Spawns (#54383)

Cenrus committed:
Dynamic configuration file.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Cenrus committed:
Allow higher accuracy requirements/antag caps.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Cenrus committed:
remove DYNAMIC_CONFIG_FILE

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Cenrus committed:
Dynamic ruleset threat scaling and various fixes (#47100)

* baby steps.

* baby steps

* Nothing works yet.

* getting closer

* Base modes work

* Squashed some bugs, fixed config.

* small cleanup.

* dynamic.json, spaces to tabs

* Disabled events on dynamic.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Third and second ruleset probability.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Highlander ruleset stacking works correctly now.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Fixed a comment.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Allow changing variables in game mode with config.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Continue after invalid variable.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* Changed how revolution ruleset ends the round.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

* fixed rule probs, stacking/ready bug, logging.

* More logging.

* Moved persistent check.

* Revs dont end on rev loss.

* Brothers count correctly no.

* Revs no longer makes borgs, deads or off-z players revheads.

* No headrev replacements.

* revs now refund on fail, reviews

* comment

* no game option changes.

Cenrus committed:
Fixes dynamic roundstart() runtime, adds more logging. (#47341)

* Fixes roundstart() runtime, adds more logging.

* one more log line.

* And fixed those lines.

Cenrus committed:
Dynamic revolution fixes, delayed ruleset improvements. (#47491)

* Revolution fixes, midround/latejoin delay, in-process refund

No more warnings for antag_cap-less modes

makes revs work again

scaling exits earlier of not threat left

initializes antag_cap again.

Fixed flavor text.

more revs stuff.

* comments, fixes

* threat log, typo.

* moved all ruleset execution to callbacks

* Clarified comment

* Status begone.

* Fixed roundend, farewell, exiles.

* team proc

* remove proc

Cenrus committed:
Fixed some dynamic bugs (#47763)

Cenrus committed:
yes

Cenrus committed:
no longer nessecary

Crossedfall commented:
Can you please link to the PRs you're cherry-picking from? Makes review a bit easier/faster

Cenrus committed:
Adds Abductors to Dynamic Spawns (#54345)

This PR does what it says right on the tin and adds Abductors as a possible choice for Dynamic mode's midround spawning. There is nothing else in this PR.

Cenrus committed:
revert dragons as they don't work

Cenrus commented:
Abductors have been implemented into midround spawns

Cenrus committed:
add clockcult roundstart

Cenrus committed:
Adds cost to cult mode

Cenrus committed:
Adds required jobs support to gamemodes (#44026)

* Adds required_jobs support to gamemodes

* alos needs to be checked with no players present

* early return

Cenrus committed:
unused var removal

PowerfulBacon commented:
Works good

PowerfulBacon commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -645,3 +672,75 @@
+	required_candidates = 4
+	weight = 3
+	cost = 35
+	requirements = list(101,101,101,101,101,101,101,101,101,101)
```
Won't this never run?

PowerfulBacon commented on config/game_options.txt:
```diff
@@ -289,6 +289,9 @@ DEFAULT_REP_VALUE 5
 ## Uncomment to allow players to see the set odds of different rounds in secret/random in the get server revision screen. This will NOT tell the current roundtype.
 #SHOW_GAME_TYPE_ODDS
 
+## Uncomment to enable dynamic ruleset config file.
```
Change config for Sage too

Cenrus committed:
oversights

Cenrus committed:
add to rotation

PowerfulBacon commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -687,8 +687,8 @@
 	cost = 35
-	requirements = list(101,101,101,101,101,101,101,101,101,101)
-	high_population_requirement = 101
+	requirements = list(2,3,3,4,4,4,4,4,4,4)
```
If I am not mistaken I believe this is the required cost for each population level (The other ones seem to be in numbers like 60,50,40,30 rather than 2,3,4,5 etc.)

Cenrus replied:
Should have checked before I changed it, fixed now

Cenrus committed:
🤦‍♀️

Crossedfall reviewed (changes requested):
Needs a copy of the `dynamic.json` file in the Sage config folder.

Cenrus committed:
final?

Cenrus commented:
Good catch, it is there now

Crossedfall commented:
I'll mirror what I asked in Discord.. To confirm, the default values are what we want for production? It seems like the config provided in the original PR was just an example, so I'm a bit unsure if it is meant for live use or not.

Crossedfall reviewed (approved):
Confirmed in Discord that this is the config we'll be sticking with for the time being.
